7th tutorial - Nilearn multilevel (two-level) glm

yibeichan opened the pull request:
Hello team!
Finally I can create PR for this 7th tutorial. This is a two-level nilearn glm, using the data from [Balloon Analog Risk-taking Task](https://openneuro.org/datasets/ds000001/versions/1.0.0), which has 16 subjects. I randomly chose 5 for our analysis. More details are in the notebook.

Here is a summary about the issues we encountered in the past month. 
After endless debugging w/ @djarecka, we found that pydra is restrict with its task input types.
1. if the input belongs to basic types such as `list`, `dict`, `str`, `int`, etc., works fine.
2. if the input is from pydra defined class such as `File`, fine too
3. if the input is a `pandas.DataFram` or `FirstLevelModel` and we define it as `ty.Any`, still okay. 
4. if the input is a nested structure of basic types such as `list[dict]` or `list[str]`, still okay.
5. HOWEVER, if the input is a list of unidentifiable types for pydra such as a list of  `pandas.DataFrame` or `FirstLevelModel`, **it will cause connection problems in the workflow that the output can't be collected or passed to the next node**. 

In this twolevel glm, the inputs for the secondlevel glm are lists. For example, that problematic node `secondlevel_estimation` needs the list of `FirstLevelModel` as inputs. More details about this node are in this issue #33 

I don't know how to solve this problem yet, but I found a way to avoid it. If I need a list of `pandas.DataFrame` as input, I saved those dataframes into file then use the list of filepath (`list[str]`). Same for `FirstLevelModel`, I saved the `firstlevel z_maps` into files and then read those file paths.

This 7th tutorial notebook runs successfully on my laptop. Hope it can pass the test here.
Thanks to @djarecka and @htwangtw for help!!

satra commented:
> HOWEVER, if the input is a list of unidentifiable types for pydra such as a list of pandas.DataFrame or FirstLevelModel, it will cause connection problems in the workflow that the output can't be collected or passed to the next node.

please open an issue in pydra regarding this. in pydra-ml we pass other non-standard objects (for example scikit-learn pipelines: https://github.com/nipype/pydra-ml/blob/master/pydra_ml/tasks.py#L103 - hence there must be something specific that's at work here)

djarecka commented:
I think that adding an extra `elif` block for input that is a data frame helped, at least in my simple example that I was using for testing. Will check for entire workflow later. 

Still has no idea why I saw some issues only when running with jupyter notebook...

djarecka commented:
@yibeichan - have you had a chance to check this failing tests? Looks like GA doesn't like the workflow for any version of python.. :(

yibeichan commented:
@djarecka yes, the problem here is with `datalad`, which needs `git-annex`. So @effigies and I create this PR #36 trying to fix the workflow by switching the set-up from python to miniconda

djarecka commented:
ok, I see it now, let's see if that helps. I'm still not able to run the workflow, but would be interested in seeing if that works on GS

djarecka commented:
@yibeichan - could you please merge the master, so the git-annex issue should be solved in the tests

effigies commented:
@djarecka The git-annex issues are resolved. The current issue is an out-of-memory failure.

djarecka commented:
@effigies - oh, ok... why do you think it is an out-of-memory failure?

effigies commented:
Because running it locally, I get `BrokenProcessPool` errors unless I close a bunch of apps and run it serially.

djarecka commented:
@effigies - is this something what you would expect with this task/data?

effigies commented:
Fitting a first-level GLM is expensive. It's also set up to fit three runs in a single model, which I believe means it's probably keeping excess things in memory. This could be split to do something more like what FitLins does, which is calculate the three models as L1 models, and calculate the fixed effects combination as a separate node.

yibeichan commented:
@effigies I tried nproc=1, 2, 4, they all failed... what else might be different between the online environment and the local environment?

djarecka commented:
I think your local machine can have more memory than we get from Github Action...

djarecka commented:
I checked on my laptop that each node can use ~5GB, so I believe it should be still fine on GA: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners
but it is pretty close to the limit

btw. all the recent changes to the notebooks seems to solve my issues on my laptop...

djarecka commented:
I've tried to run GA with `ubuntu-20.04-16core` (https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners), but looks like waiting for the runner takes forever...
https://github.com/djarecka/pydra-tutorial/pull/2

yibeichan commented:
> I checked on my laptop that each node can use ~5GB, so I believe it should be still fine on GA: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners but it is pretty close to the limit
> 
> btw. all the recent changes to the notebooks seems to solve my issues on my laptop...

so the good news is that the current version works for all of us (OS: MacOX, Linux; Python: 3.7, 3.8, 3.10) locally

> I've tried to run GA with `ubuntu-20.04-16core` (https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners), but looks like waiting for the runner takes forever... [djarecka#2](https://github.com/djarecka/pydra-tutorial/pull/2)

I just checked yours, it's still queued (5 hours after you set it up)...

I am at my wit's end.... hahahah(with a wry face)

yibeichan commented:
@djarecka we have the same error here `AttributeError: module 'pydra' has no attribute '__version__'` as in pydra-fsl tests
https://github.com/nipype/pydra-fitlins/actions/runs/3034042858/jobs/4882728928#step:5:58
we didn't have this error last month. I checked that the recent successful devcheck was on Aug 3rd ([here](https://github.com/nipype/pydra-fsl/actions/runs/2793846708/jobs/4401929044)). Did we change anything in Pydra after that?

yibeichan commented:
hello @djarecka @effigies I rewrote the first level GLM as we discussed - separate the firstlevel model and fixed-effect. The notebook works fine on my Mac (python 3.7 & 3.10). However, it still fails GHA. Maybe it's not a memory issue?

djarecka commented:
Thank you! I've just run the workflow on my laptop and I could see that sometimes memory was around 7GB, so it definitely could be still a memory issue..:( Perhaps we should just downsample??

yibeichan commented:
hello Team! Finally, after one month, this notebook passed GHA. Yes, it was the memory issue. Thanks Dorota helping me find the solution for down sampling, which was a solution Chris provided on Neurostars :). @effigies once you get time, please check and merge. Thank you!

satra commented on notebooks/6_firstlevel_glm_nilearn.md:
````diff
@@ -525,7 +510,7 @@ wf.set_output(
 
-```{code-cell}
+```{code-cell} ipython3
 :tags: []
````
```suggestion
:tags: [hide-output]
```

yibeichan replied:
Thanks @satra! I added it, also made input/output hidden for the 7th notebook

yibeichan commented:
I made a few formatting/structural changes to have the jupyter book look prettier. 
Then, I added one plot (unthresholded) for the second level GLM in the 7th notebook to prove that all other plots having nothing survived is due to the small sample size (n=5) not that things are wrong with the workflow.

djarecka commented:
Thank you! Great job!

djarecka commented:
@effigies @htwangtw - I've merged it, but if you have any comments, you can open an issue and we can fix it in a separate pr!
